site/book/05-developing-functions/02-developing-in-Go.md

```diff
@@ -1,133 +1,177 @@
-			if err != nil {
-				return nil, err
-			}
-		}
-		return items, nil
-	}
-	p := framework.SimpleProcessor{Filter: kio.FilterFunc(fn), Config: &config}
-	cmd := command.Build(p, command.StandaloneDisabled, false)
-  // Adds a "gen" subcommand to create a Dockerfile for building the function into a container image.
-  command.AddGenerateDockerfile(cmd)
-
-	if err := cmd.Execute(); err != nil {
-		os.Exit(1)
-	}
+  // CUSTOMIZE IF NEEDED
+  // `AsMain` accepts a `ResourceListProcessor` interface.
+  // You can explore other `ResourceListProcessor` structs in the SDK or define your own.
+  if err := fn.AsMain(fn.ResourceListProcessorFunc(Run)); err != nil {
+    os.Exit(1)
+  }
 }
 ```
 
-### Build and test the function
+The [`fn`] library provides a series of KRM level operations for [`ResourceList`]. 
+Basically, the KRM resource `ResourceList.FunctionConfig` and KRM resources `ResourceList.Items` are both converted to 
+`KubeObject` objects. You can use `KubeObject` similar as [`unstructured.Unstrucutred`].
 
-Build the go binary:
+The set-annotation function (see below) iterates the `ResourceList.Items`, finds out the `Deployment` resources and
+adds the annotation. After the iteration, it adds some user message to the `ResourceList.Results`
 
-```shell
-$ go mod tidy; go build -o my-fn .
+```go
+func Run(rl *fn.ResourceList) (bool, error) {
+    for _, kubeObject := range rl.Items {
+        if kubeObject.IsGVK("v1", "Deployment") {
+            kubeObject.SetAnnotation("config.kubernetes.io/managed-by", "kpt")
+        }
+    }
+    rl.Results = append(rl.Results, fn.GeneralResult("Add config.kubernetes.io/managed-by=kpt to all `Deployment` resources", fn.Info))
+    return true, nil
+}
 ```
 
-Fetch the wordpress package:
+Learn more about the `KubeObject` from the [go doc](https://pkg.go.dev/github.com/GoogleContainerTools/kpt-functions-sdk/go/fn)
 
-```shell
-$ kpt pkg get https://github.com/GoogleContainerTools/kpt.git/package-examples/wordpress@v0.8
-```
 
-Test it by running the function imperatively:
+### Test the KRM function
+
+The "get-started" package contains a `./data` directory. You can use this to test out your functions. 
 
 ```shell
-$ kpt fn eval wordpress --exec ./my-fn -- myannotation=foo
-```
+# Edit the `data/resources.yaml` with your KRM resources. 
+# resources.yaml already has a `Deployment` and `Service` as test data. 
+vim data/resources.yaml
 
-During iterative development, `--exec` flag can be used to execute the
-function binary directly instead of requiring the function to be containerized
-first. Once you have a function binary that works, you can then proceed to
-creating the container image.
+# Convert the KRM resources and FunctionConfig resource to `ResourceList`.
+kpt fn source data
+
+# Pass the ResourceList as StdIn to your function
+kpt fn source data | go run main.go
 
-### Publish the function
+# Verify the KRM function behavior in the StdOutput `ResourceList`
+```
```

Comment:
One thing I observed: we can reverse the order here:

First test the function:
```
kpt fn source data| go run main.go
```
See if all `deployments` have the new label or not ?

Now talk about tweaking the inputs under `data` directory.

Then... if you are curious about the input `ResourceList` to the function, here is how you can see it

```
kpt fn source data
```

Reply:
sounds good. I simplified these two steps but give more explanations.